src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryDataLoader.kt

```diff
@@ -408,6 +439,217 @@
         list.add(upgrade)
     }
 
+    private fun findAllBestUpgrades(list: List<ChocolateFactoryUpgrade>) {
+
+        val curChocolate = ChocolateAmount.CURRENT.chocolate() + ChocolateAmount.chocolateSinceUpdate()
+
+        // Only look at upgrades that would increase the average chocolate per second.
+        // removing time tower here as people like to determine when to buy it themselves.
+        val currentUpgrades = ArrayList(
+            list.filter { it.extraPerSecond != null && it.extraPerSecond > 0 }
+                .filter { it.slotIndex != ChocolateFactoryAPI.timeTowerIndex }
+        )
+
+        // check if any of the current upgrades is of higher level then previous calculated best upgrades
+        val isCurrentHigher = currentUpgrades.any { current ->
+            val upgrade = ChocolateFactoryAPI.allBestPossibleUpgrades[current.slotIndex]?.lastOrNull()
+            val lvlOnLastRun = ChocolateFactoryAPI.lastUpgradesWhenChecking[current.slotIndex]
+            return@any /*Last level was not set (e.g., first run after clear)*/lvlOnLastRun == null
```

Comment:
pleasedont use in between comments. rather extract variables with descriptive names